platforms/posix/src/px4/common/px4_daemon/server.cpp

```diff
@@ -210,10 +213,8 @@ Server::_server_main()
 				auto thread = _fd_to_thread.find(poll_fds[i].fd);
 
 				if (thread != _fd_to_thread.end()) {
-					// Thread is still running, so we cancel it.
-					// TODO: use a more graceful exit method to avoid resource leaks
-					pthread_cancel(thread->second);
-					_fd_to_thread.erase(thread);
+					// Thread is still running, so we send a byte to stop_thread.
+					ret = write(_fd_to_stop_thread[poll_fds[i].fd][1], "x", 1);
```

Comment:
Unlike `pthread_cancel`, pipe can only take effect before read is called.  

When any one of the pipe or socket receives data, the blocking of `select` will end, which avoids the situation that the thread is blocked in the `read` function and cannot respond to the stop_thread pipe.

Reply:
> Unlike pthread_cancel, pipe can only take effect before read is called.

I'm not quite sure what you mean but a write to a pipe is non-blocking, there's a buffer.

> When any one of the pipe or socket receives data, the blocking of select will end, which avoids the situation that the thread is blocked in the read function and cannot respond to the stop_thread pipe.

That assumes the thread is waiting in the select, but this is not necessarily true. It could already be inside the `read` (from a previous wake-up).

Reply:
The process here is as follows:
1. The socket has incoming data, `select` ends blocking
2. `read` has been called
3. `read` reads all the data
4. The next round of loop runs to `select` and continues to block.

We don't need to worry about the thread could already be inside the `read`, because `select` ensures that the socket has data incoming, and `read` will definitely end in a short time.

Reply:
> and read will definitely end in a short time.

That's not the proper way to solve or think about race conditions. Please fix it completely.